lib/rust/ensogl/core/src/display/object.rs

```diff
@@ -23,3 +28,138 @@ pub mod traits {
 }
+
+
+
+// =============================
+// === InstanceWithSublayers ===
+// =============================
+
+/// A wrapper around [`Instance`] with attached sublayers. It is used for
+/// dynamic creation and managing of additional on-demand sublayers needed for various UI
+/// components.
+///
+/// When the Instance is added to another layer, [`Sublayers::remove_from_layer`] and
+/// [`Sublayers::add_to_layer`] methods of the `sublayers` field will be called with previous and
+/// new layers respectively.
+///
+/// See [`MaskingSublayers`] docs for one of the use cases.
+#[derive(Debug, Clone, CloneRef)]
+pub struct InstanceWithSublayers<S: CloneRef> {
+    inner:         Instance,
+    /// Attached sublayers. Should implement [`Sublayers`] trait.
+    pub sublayers: S,
+}
+
+impl<S: CloneRef> Deref for InstanceWithSublayers<S> {
+    type Target = Instance;
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<S: CloneRef + Sublayers + 'static> InstanceWithSublayers<S> {
+    /// Constructor. Sets a `on_scene_layers_changed` callback for `instance`.
+    pub fn new(instance: Instance, sublayers: S) -> Self {
+        instance.set_on_scene_layer_changed(f!([sublayers](_, from, to) {
+            for layer in from {
+                if let Some(layer) = layer.upgrade() {
+                    sublayers.remove_from_layer(&layer);
+                }
+            }
+            for layer in to {
+                if let Some(layer) = layer.upgrade() {
+                    sublayers.add_to_layer(&layer);
+                }
+            }
+        }));
+        Self { inner: instance, sublayers }
+    }
+}
+
+
+
+// ========================
+// === MaskingSublayers ===
+// ========================
+
+/// Sublayers for easier masking with arbitrary shapes.
+///
+/// One of the examples might be a `ScrollArea` component implementation. To clip the area's content
+/// (so that it is displayed only inside its borders) we use `MaskingSublayers` with two additional
+/// layers: one for the area's content and another one for the masking shape. These sublayers will
+/// automatically follow the main [`Instance`] if it moves between different
+/// layers and will be dropped together with the main object.
+#[derive(Debug, Clone, CloneRef)]
+pub struct MaskingSublayers {
+    /// Contains the objects that needs to be masked.
+    pub content: Layer,
+    /// Contains a masking shape. See docs in [`crate::display::scene::layer`] for information on
+    /// masking.
+    pub mask:    Layer,
+}
+
+impl MaskingSublayers {
+    /// Constructor. The passed `camera` will be used to render created sublayers.
+    pub fn new(logger: &Logger, camera: &Camera2d) -> Self {
+        let content = Layer::new_with_cam(logger.sub("ContentLayer"), camera);
+        let mask = Layer::new_with_cam(logger.sub("MaskLayer"), camera);
+        content.set_mask(&mask);
+        Self { content, mask }
+    }
+}
```

Comment:
This abstraction does not seem like a display object one. It should rather be placed in layers instead.